src/com/magento/idea/magento2plugin/completion/xml/XmlCompletionContributor.java

```diff
@@ -18,22 +21,46 @@
 public class XmlCompletionContributor extends CompletionContributor {
 
     public XmlCompletionContributor() {
-        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_ATTRIBUTE_VALUE_TOKEN),
-            new CompositeCompletionProvider(
-                new PhpClassCompletionProvider(),
-                new PhpClassMemberCompletionProvider(),
-                new ModuleNameCompletionProvider(),
-                new FilePathCompletionProvider()
-            )
+
+        /* PHP class member completion provider */
+        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_DATA_CHARACTERS)
+                        .withParent(XmlPatterns.xmlText().withParent(XmlPatterns.xmlTag().withChild(
+                                XmlPatterns.xmlAttribute().withName(CommonXml.SCHEMA_VALIDATE_ATTRIBUTE)
+                                        .withValue(string().oneOf(CommonXml.INIT_PARAMETER))))
+                        ),
+                new PhpClassMemberCompletionProvider()
+        );
+
+        /* Module Completion provider */
+        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_ATTRIBUTE_VALUE_TOKEN)
+                        .inside(XmlPatterns.xmlAttribute().withName(ModuleAclXml.XML_ATTR_ID))
+                        .inFile(xmlFile().withName(string().endsWith(ModuleAclXml.FILE_NAME))),
+                new ModuleNameCompletionProvider()
+        );
+
+        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_ATTRIBUTE_VALUE_TOKEN)
+                        .inside(XmlPatterns.xmlAttribute().withName(ModuleXml.MODULE_ATTR_NAME))
+                        .inFile(xmlFile().withName(string().endsWith(ModuleXml.FILE_NAME))),
+                new ModuleNameCompletionProvider()
+        );
+
+        /* PHP Class completion provider */
+        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_DATA_CHARACTERS)
+                        .withParent(XmlPatterns.xmlText().withParent(XmlPatterns.xmlTag().withChild(
+                                XmlPatterns.xmlAttribute().withName(CommonXml.SCHEMA_VALIDATE_ATTRIBUTE).withValue(string().oneOf(CommonXml.OBJECT))))
+                        ),
+                new PhpClassCompletionProvider()
         );
 
-        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_DATA_CHARACTERS),
-            new CompositeCompletionProvider(
-                new PhpClassCompletionProvider(),
-                new PhpClassMemberCompletionProvider(),
-                new ModuleNameCompletionProvider(),
+        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_ATTRIBUTE_VALUE_TOKEN)
+                        .inside(XmlPatterns.xmlAttribute().withName(CommonXml.ATTR_CLASS)),
+                new PhpClassCompletionProvider()
+        );
+
+        /* File Path Completion provider */
+        extend(CompletionType.BASIC, psiElement(XmlTokenType.XML_ATTRIBUTE_VALUE_TOKEN)
+                        .inside(XmlPatterns.xmlAttribute().withName("template")),
```

Comment:
Hi.
I've refactored most of the string literals to constant classes. Please do the same way.
 Thank you!

Reply:
We are storing all magento related string literals under `com/magento/idea/magento2plugin/magento`.

Reply:
I'm about the new changes.

Reply:
Okay, I will make sure of these changes